### ISSUE TITLE
Argument of type "X" cannot be assigned to parameter of type "X" despite cast

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#9953]
**Describe the bug**
I would expect that after `array = cast("FloatArray", array)`, I should always be able to call `foo(array)` with `def foo(array: FloatArray)`. However, `pyright` does not like it:

> bug.py:20:9 - error: Argument of type "FloatArray" cannot be assigned to parameter "array" of type "FloatArray" in function "foo"
    "ndarray[Any, dtype[floating[Any] | integer[Any]]]" is not assignable to "ndarray[Any, dtype[floating[Any]]]"
      Type parameter "_DType_co@ndarray" is covariant, but "dtype[floating[Any] | integer[Any]]" is not a subtype of "dtype[floating[Any]]"
        "dtype[floating[Any] | integer[Any]]" is not assignable to "dtype[floating[Any]]"
          Type parameter "_DTypeScalar_co@dtype" is covariant, but "floating[Any] | integer[Any]" is not a subtype of "floating[Any]"
            Type "floating[Any] | integer[Any]" is not assignable to type "floating[Any]" (reportArgumentType)
1 error, 0 warnings, 0 informations

**Code or Screenshots**
```python
from typing import Any, cast

import numpy as np
import numpy.typing as npt

type FloatArray = npt.NDArray[np.floating[Any]]
type NumberArray = npt.NDArray[np.floating[Any] | np.integer[Any]]


def foo(array: FloatArray) -> None: ...


def bar(array: NumberArray) -> None:
    # this works:
    array2 = cast("FloatArray", array)
    foo(array2)

    # this seems to confuse pyright
    array = cast("FloatArray", array)
    foo(array)
```

**VS Code extension or command-line**
pyright 1.1.394 from PyPI

**Comment by decorator-factory** [comment]
To my understanding, Pyright always treats a parameter as an "explicitly annotated variable".

For example, pyright allows this:
```py
message = b"hello, world"   # message is `bytes` after this line
message = message.decode()  # message is `str` after this line
```
But not this: 
```py
message: bytes = b"hello, world"
message = message.decode()  # error: Type "str" is not assignable to declared type "bytes"
```
And arguments are treated in the same way:
```py
def f(message: bytes) -> None:
    message = message.decode()  # error: Type "str" is not assignable to declared type "bytes"
```

**Comment by erictraut** [comment]
Addressing this issue required me to rethink the algorithm pyright uses for narrowing types on assignment — in particular, in cases where either the target (LHS) declared type or the assigned (RHS) type contains an `Any`. I've known for some time that pyright's algorithm here was inconsistent. This bug highlighted the fact that under specific circumstances it produced nonsensical results. 

I've reimplemented the algorithm so it is simpler and consistent — and more grounded in type theory. The new behavior more closely matches that of mypy, which is also a good thing.

The downside to this change is that it might have some backward compatibility impact for code bases that use pyright as a type checker — especially if "strict" type checking is used. The differences should be minor, but they may require some small changes in certain cases.

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.398.
